File: backend_flask_small/routes/owners.py
from flask import Blueprint, request, jsonify
from db import get_db_connection, release_db_connection
from auth_utils import hash_password, check_password, generate_tokens, hash_token
import datetime
import uuid

owners_bp = Blueprint('owners', __name__)

@owners_bp.route('/register', methods=['POST'])
def register():
    data = request.get_json()
    email = data.get('email')
    password = data.get('password')
    full_name = data.get('full_name')
    public_key = data.get('public_key')

    if not email or not password:
        return jsonify({'error': 'email and password required'}), 400
    
    if not public_key:
        return jsonify({'error': 'public_key is required'}), 400

    hashed_password = hash_password(password)
    conn = get_db_connection()
    cursor = conn.cursor()

    try:
        cursor.execute("SELECT id FROM owners WHERE email = ?", (email,))
        if cursor.fetchone():
            return jsonify({'error': 'Owner already exists'}), 409

        owner_id = str(uuid.uuid4())
        cursor.execute(
            "INSERT INTO owners (id, email, password_hash, full_name, public_key, private_key_encrypted) VALUES (?, ?, ?, ?, ?, ?)",
            (owner_id, email, hashed_password, full_name, public_key, '')
        )
        
        cursor.execute("SELECT id, email, full_name FROM owners WHERE id = ?", (owner_id,))
        owner = cursor.fetchone()
        owner_email = owner[1]
        owner_name = owner[2]

        access_token, refresh_token = generate_tokens(owner_id, owner_email, 'owner')

        session_id = str(uuid.uuid4())
        hashed_refresh = hash_token(refresh_token)
        hashed_access = hash_token(access_token)
        expires_at = datetime.datetime.utcnow() + datetime.timedelta(days=7)
        
        cursor.execute(
            """INSERT INTO sessions (id, user_id, token_hash, refresh_token_hash, expires_at, refresh_expires_at, is_valid)
               VALUES (?, ?, ?, ?, ?, ?, 1)""",
            (session_id, owner_id, hashed_access, hashed_refresh, expires_at, expires_at)
        )
        
        conn.commit()

        return jsonify({
            'success': True,
            'accessToken': access_token,
            'refreshToken': refresh_token,
            'owner': {'id': owner_id, 'email': owner_email, 'full_name': owner_name}
        }), 201

    except Exception as e:
        conn.rollback()
        print(f"Owner Register error: {e}")
        return jsonify({'error': True, 'message': 'Registration failed'}), 500
    finally:
        cursor.close()
        release_db_connection(conn)

@owners_bp.route('/login', methods=['POST'])
def login():
    data = request.get_json()
    email = data.get('email')
    password = data.get('password')

    if not email or not password:
        return jsonify({'error': 'email and password required'}), 400

    conn = get_db_connection()
    cursor = conn.cursor()

    try:
        cursor.execute("SELECT id, email, password_hash, full_name FROM owners WHERE email = ?", (email,))
        owner = cursor.fetchone()

        if not owner or not check_password(password, owner[2]):
            return jsonify({'error': 'Invalid credentials'}), 401

        owner_id = owner[0]
        owner_email = owner[1]
        owner_name = owner[3]

        access_token, refresh_token = generate_tokens(owner_id, owner_email, 'owner')

        session_id = str(uuid.uuid4())
        hashed_refresh = hash_token(refresh_token)
        hashed_access = hash_token(access_token)
        expires_at = datetime.datetime.utcnow() + datetime.timedelta(days=7)

        cursor.execute(
            """INSERT INTO sessions (id, user_id, token_hash, refresh_token_hash, expires_at, refresh_expires_at, is_valid)
               VALUES (?, ?, ?, ?, ?, ?, 1)""",
            (session_id, owner_id, hashed_access, hashed_refresh, expires_at, expires_at)
        )
        conn.commit()

        return jsonify({
            'success': True,
            'accessToken': access_token,
            'refreshToken': refresh_token,
            'owner': {'id': owner_id, 'email': owner_email, 'full_name': owner_name}
        })

    except Exception as e:
        conn.rollback()
        print(f"Owner Login error: {e}")
        return jsonify({'error': True, 'message': 'Login failed'}), 500
    finally:
        cursor.close()
        release_db_connection(conn)

@owners_bp.route('/public-key/<identifier>', methods=['GET'])
def get_public_key(identifier):
    conn = get_db_connection()
    cursor = conn.cursor()
    try:
        # Check if identifier looks like an email
        if '@' in identifier:
            cursor.execute("SELECT public_key FROM owners WHERE email = ?", (identifier,))
        else:
            # Assume UUID
            cursor.execute("SELECT public_key FROM owners WHERE id = ?", (identifier,))
            
        result = cursor.fetchone()
        if not result:
            return jsonify({'error': 'Owner not found'}), 404
        
        return jsonify({'success': True, 'publicKey': result[0]})
        
    except Exception as e:
        print(f"Get Public Key error: {e}")
        return jsonify({'error': True, 'message': 'Failed to fetch key'}), 500
    finally:
        cursor.close()
        release_db_connection(conn)

File: backend_docker_large/backend/routes/owners.js
const express = require("express");
const router = express.Router();
const db = require("../database");
const AuthService = require("../services/authService");
const { verifyToken, verifyRole } = require("../middleware/auth");

// ========================================
// SECURITY NOTE: Private Key Handling
// ========================================
// Private keys are NEVER generated or stored on the server.
// Owners MUST generate keypairs client-side and securely store private keys locally.
// The server only accepts and stores public keys.
// This prevents accidental exposure of private keys via logs, responses, or backups.
// ========================================

// POST /api/owners/register
// Requires client to generate RSA keypair and submit public key
router.post("/register", async (req, res) => {
  try {
    const { email, password, full_name, public_key } = req.body;

    // Validate required fields
    if (!email || !password) {
      return res.status(400).json({ error: "email and password required" });
    }
    if (!public_key) {
      return res.status(400).json({
        error: "public_key is required",
        message:
          "You must generate an RSA-2048 keypair client-side and submit the public key (PEM format)",
      });
    }

    if (!AuthService.validateEmail(email)) {
      return res.status(400).json({ error: "Invalid email format" });
    }

    // Validate public key format (basic check for PEM format)
    if (!public_key.startsWith("-----BEGIN PUBLIC KEY-----")) {
      return res.status(400).json({
        error: "invalid_public_key_format",
        message: "Public key must be in PEM format (SPKI)",
      });
    }

    const exists = await db.query("SELECT id FROM owners WHERE email = $1", [
      email,
    ]);
    if (exists.rows.length > 0) {
      return res.status(409).json({ error: "Owner already exists" });
    }

    const passwordHash = await AuthService.hashPassword(password);

    const insertQ = `
      INSERT INTO owners (email, password_hash, full_name, public_key)
      VALUES ($1, $2, $3, $4)
      RETURNING id, email, full_name, created_at
    `;

    const result = await db.query(insertQ, [
      email,
      passwordHash,
      full_name || null,
      public_key,
    ]);
    const owner = result.rows[0];

    const payload = { sub: owner.id, email: owner.email, role: "owner" };
    const accessToken = AuthService.generateAccessToken(payload);
    const refreshToken = AuthService.generateRefreshToken(payload);

    // Store session
    const hashedRefresh = AuthService.hashToken(refreshToken);
    const expiresAt = new Date(Date.now() + 7 * 24 * 3600 * 1000);
    await db.query(
      `INSERT INTO sessions (owner_id, token_hash, refresh_token_hash, expires_at, refresh_expires_at, is_valid)
       VALUES ($1, $2, $3, $4, $4, true)`,
      [owner.id, AuthService.hashToken(accessToken), hashedRefresh, expiresAt]
    );

    // SECURITY: Do NOT return any private key material
    res.status(201).json({
      success: true,
      message: "Owner registered successfully",
      accessToken,
      refreshToken,
      owner: { id: owner.id, email: owner.email, full_name: owner.full_name },
      note: "Ensure your private key is stored securely on your device. Keep it safe and never share it.",    });
  } catch (error) {
    const isDev = process.env.NODE_ENV === "development";
    const errorMsg = `Owner registration failed: ${error.code || error.name || 'UNKNOWN'}`;
    if (isDev) console.error(errorMsg, { stack: error.stack });
    else console.error(errorMsg);
    res.status(500).json({
      error: true,
      message: "Owner registration failed",
      ...(isDev && { details: error.message }),
    });
  }
});

// POST /api/owners/login
router.post("/login", async (req, res) => {
  try {
    const { email, password } = req.body;
    if (!email || !password) {
      return res.status(400).json({ error: "email and password required" });
    }

    const result = await db.query(
      "SELECT id, email, password_hash, full_name FROM owners WHERE email = $1",
      [email]
    );
    if (result.rows.length === 0) {
      return res.status(401).json({ error: "Invalid credentials" });
    }

    const owner = result.rows[0];
    const ok = await AuthService.comparePassword(password, owner.password_hash);
    if (!ok) {
      return res.status(401).json({ error: "Invalid credentials" });
    }

    const payload = { sub: owner.id, email: owner.email, role: "owner" };
    const accessToken = AuthService.generateAccessToken(payload);
    const refreshToken = AuthService.generateRefreshToken(payload);

    const hashedRefresh = AuthService.hashToken(refreshToken);
    const expiresAt = new Date(Date.now() + 7 * 24 * 3600 * 1000);
    await db.query(
      `INSERT INTO sessions (owner_id, token_hash, refresh_token_hash, expires_at, refresh_expires_at, is_valid)
       VALUES ($1, $2, $3, $4, $4, true)`,
      [owner.id, AuthService.hashToken(accessToken), hashedRefresh, expiresAt]
    );

    res.json({
      success: true,
      accessToken,
      refreshToken,
      owner: { id: owner.id, email: owner.email, full_name: owner.full_name },
    });
  } catch (error) {
    const isDev = process.env.NODE_ENV === "development";
    const errorMsg = `Owner login failed: ${error.code || error.name || 'UNKNOWN'}`;
    if (isDev) console.error(errorMsg, { stack: error.stack });
    else console.error(errorMsg);
    res.status(500).json({
      error: true,
      message: "Owner login failed",
      ...(isDev && { details: error.message }),
    });
  }
});

// GET /api/owners/public-key/:ownerId
// Retrieve public key for a specific owner (no auth required - keys are public)
router.get("/public-key/:ownerId", async (req, res) => {
  try {
    const { ownerId } = req.params;

    // Validate ownerId format
    if (!ownerId || ownerId.length < 5) {
      return res.status(400).json({ error: "Invalid owner ID" });
    }

    const result = await db.query(
      "SELECT public_key FROM owners WHERE id = $1",
      [ownerId]
    );
    if (result.rows.length === 0) {
      return res.status(404).json({ error: "Owner not found" });
    }

    res.json({ success: true, public_key: result.rows[0].public_key });
  } catch (error) {
    const isDev = process.env.NODE_ENV === "development";
    const errorMsg = `Failed to fetch public key: ${error.code || error.name || 'UNKNOWN'}`;
    if (isDev) console.error(errorMsg, { stack: error.stack });
    else console.error(errorMsg);
    res.status(500).json({
      error: true,
      message: "Failed to fetch public key",
      ...(isDev && { details: error.message }),
    });
  }
});

// GET /api/owners/me (requires authentication)
// Get authenticated owner's profile (does NOT return private key)
router.get("/me", verifyToken, verifyRole(["owner"]), async (req, res) => {
  try {
    const ownerId = req.user.sub;

    const result = await db.query(
      "SELECT id, email, full_name, created_at FROM owners WHERE id = $1",
      [ownerId]
    );
    if (result.rows.length === 0) {
      return res.status(404).json({ error: "Owner not found" });
    }

    res.json({ success: true, owner: result.rows[0] });
  } catch (error) {
    const isDev = process.env.NODE_ENV === "development";
    const errorMsg = `Failed to fetch owner profile: ${error.code || error.name || 'UNKNOWN'}`;
    if (isDev) console.error(errorMsg, { stack: error.stack });
    else console.error(errorMsg);
    res.status(500).json({
      error: true,
      message: "Failed to fetch owner profile",
      ...(isDev && { details: error.message }),
    });
  }
});

module.exports = router;
